File: Base Classes/BaseActivity.java
package com.example.myapplication;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.drawable.Drawable;
import android.net.ConnectivityManager;
import android.os.Bundle;
import androidx.annotation.ColorInt;
import androidx.annotation.ColorRes;
import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

public abstract class BaseActivity extends AppCompatActivity  {

    public static final String TAG = "BaseActivity.";
    //  @getContentView() Your layout
    protected abstract int getContentView();
    //  @initViews() Initialize your views and start code
    protected abstract void initViews();
    //  Interceptor for catch internet status
    private final BroadcastReceiver internetReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            boolean noConnect = intent.getBooleanExtra(ConnectivityManager.EXTRA_NO_CONNECTIVITY, false);
            onNetworkChangeUI(!noConnect);
        }
    };
    //  Method override for change UI if there is no internet
    public void onNetworkChangeUI(boolean isInternet) {

    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getContentView());
        initViews();
    }

    @Override
    protected void onResume() {
        super.onResume();
        //  Bind CONNECTIVITY_ACTION for check network status
        try {
            registerReceiver(internetReceiver, new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
        } catch (Exception e) {
            //  Your code
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        //  Unbind CONNECTIVITY_ACTION
        try {
            unregisterReceiver(internetReceiver);
        } catch (Exception e) {
            //  Your code
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        onEnd();
    }

    //  Method override, catch destroy activity
    protected void onEnd() {

    }
    //  Resources ID
    @NonNull
    public String getStr(@StringRes int id) {
        return getString(id);
    }

    @NonNull
    public String concatStr(String text) {
        return text;
    }

    @Nullable
    public Drawable getImg(@DrawableRes int id) {
        return ContextCompat.getDrawable(this, id);
    }

    @ColorInt
    public int getClr(@ColorRes int id) {
        return ContextCompat.getColor(this, id);
    }
}

File: Base Classes/BaseActivity.kt
package com.example.myapplication

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.drawable.Drawable
import android.net.ConnectivityManager
import android.os.Bundle
import androidx.annotation.*
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat

abstract class BaseActivity(@LayoutRes private val layoutResID: Int): AppCompatActivity() {

    companion object {
        const val TAG = "BaseActivity"
    }
    //  Initialize your views and start code
    protected abstract fun initViews()
    //  Interceptor for catch internet status
    private val internetReceiver = object: BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            val noConnect = intent?.getBooleanExtra(ConnectivityManager.EXTRA_NO_CONNECTIVITY, false)
            onNetworkChangeUI(!noConnect!!)
        }

    }

    //  Method override for change UI if there is no internet
    open fun onNetworkChangeUI(isInternet: Boolean) {

    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //  Before UI create
        onBegin()
        //  Override Resources ID Layout
        setContentView(layoutResID)
        //  After UI create and initialize all views inside this layout and access all widgets by ID
        initViews()
    }

    override fun onResume() {
        super.onResume()
        //  Bind CONNECTIVITY_ACTION for check network status
        try {
            registerReceiver(internetReceiver, IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
        } catch (e: Exception) {
            //  Your code
        }
    }

    override fun onPause() {
        super.onPause()
        //  Unbind CONNECTIVITY_ACTION
        try {
            unregisterReceiver(internetReceiver);
        } catch (e: Exception) {
            //  Your code
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        onEnd()
    }

    //  Method override, catch create fragment
    open fun onBegin() {

    }

    //  Method override, catch destroy activity
    open fun onEnd() {

    }

    //  Resources ID
    open fun getStr(@StringRes id: Int): String {
        /*
         *  If your app supported more one language,
         *  you make add your locale
         *  example -> yourResources.getString(id);
         */
        return getString(id)
    }

    open fun concatStr(text: String): String {
        /*
         * Concat all your text, strings and resources,
         * to one String
         */
        return text
    }

    @Nullable
    open fun getImg(@DrawableRes id: Int): Drawable? {
        //  Short method name and override future deprecated methods
        return ContextCompat.getDrawable(this, id)
    }

    @ColorInt
    open fun getClr(@ColorRes id: Int): Int {
        //  Short method name and override future deprecated methods
        return ContextCompat.getColor(this, id)
    }
}
